homeassistant/components/heos/media_player.py

```diff
@@ -0,0 +1,178 @@
+"""Denon HEOS Media Player."""
+
+from homeassistant.components.media_player import MediaPlayerDevice
+from homeassistant.components.media_player.const import (
+    DOMAIN, MEDIA_TYPE_MUSIC, SUPPORT_NEXT_TRACK, SUPPORT_PAUSE, SUPPORT_PLAY,
+    SUPPORT_PLAY_MEDIA, SUPPORT_PREVIOUS_TRACK, SUPPORT_STOP,
+    SUPPORT_VOLUME_MUTE, SUPPORT_VOLUME_SET, SUPPORT_VOLUME_STEP)
+from homeassistant.const import STATE_OFF, STATE_PAUSED, STATE_PLAYING
+
+import asyncio
+import logging
+from . import DOMAIN as HEOS_DOMAIN
+
+DEPENDENCIES = ['heos']
+
+SUPPORT_HEOS = SUPPORT_PLAY | SUPPORT_STOP | SUPPORT_PAUSE | \
+        SUPPORT_PLAY_MEDIA | SUPPORT_PREVIOUS_TRACK | SUPPORT_NEXT_TRACK | \
+        SUPPORT_VOLUME_MUTE | SUPPORT_VOLUME_SET | SUPPORT_VOLUME_STEP
+
+PLAY_STATE_TO_STATE = {
+    'play': STATE_PLAYING,
+    'pause': STATE_PAUSED,
+    'stop': STATE_OFF
+}
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup_platform(hass, config, async_add_devices,
+                               discover_info=None):
+    """Set up the HEOS platform."""
+    controller = hass.data[HEOS_DOMAIN][DOMAIN]
+    try:
+        await asyncio.wait_for(
+            _heos_setup(controller, async_add_devices),
+            timeout=5.0)
+    except asyncio.TimeoutError:
+        _LOGGER.error('Timeout during setup.')
+
+
+async def _heos_setup(controller, async_add_devices):
+    """HEOS initial setup"""
+    await controller.connect()
+
+    players = controller.get_players()
+    devices = [HeosMediaPlayer(p) for p in players]
+    async_add_devices(devices, True)
+
+
+class HeosMediaPlayer(MediaPlayerDevice):
+    """The HEOS player."""
+
+    def __init__(self, player):
+        """Initialize."""
+        self._player = player
+        self._dispatcher_remove = None
+        self._player.state_change_callback = None
+
+    def _update_state(self):
+        self.async_schedule_update_ha_state()
+
+    async def async_update(self):
+        """Update the player."""
+        self._player.request_update()
+
+    async def async_added_to_hass(self):
+        """Device added to hass."""
+        self._player.state_change_callback = self._update_state
+        await self.async_update_ha_state()
```

Comment:
We don't need to update state here. It will be done as the last part of adding the entity.